torchaudio/functional.py

```diff
@@ -466,3 +468,104 @@ def phase_vocoder(complex_specgrams, rate, phase_advance):
     complex_specgrams_stretch = torch.stack([real_stretch, imag_stretch], dim=-1)
 
     return complex_specgrams_stretch
+
+
+@torch.jit.script
+def stft(waveform, pad, window, n_fft, hop_length, win_length):
```

Comment:
stft is already in pytorch core. I know that not having batching is annoying but replicating a function across both libraries is too. We're working on a more principled abstraction around batching, but we'd like to avoid having a reimplementation of this. Maybe we can do this in a separate PR?

Reply:
>We're working on a more principled abstraction around batching, but we'd like to avoid having a reimplementation of this. Maybe we can do this in a separate PR?

👍 

Gotcha, so I can take this out no problem and then have the `STFT` layer just use `torch.stft`? I think the bigger point for this was to have the non-normalized output of stft as a functional/layer.

Reply:
> ## On batching
> I've made `STFT` be able to handle a batch dimension, and so can the augmentation layers. I can't seem to find it but I remember a discussion on how the layers should handle batch of inputs. Should the layers apply transforms only to single inputs for now? I can make changes if that is the case.

In this PR, we can focus on single inputs in order to avoid adding STFT here again, as mentioned by @cpuhrsch. For batching, I would prefer waiting on a standardized approach. If that is so important, I'd open a separate PR for that so we don't block the rest :)

Reply:
Are you saying there's an operation provided by torchaudio that normalizes the output when you would like it not to?

Reply:
Yeah, so [`Spectrogram`](https://github.com/pytorch/audio/blob/8273c3f4500ac9c01e95f1d78fa3f5752aafdca7/torchaudio/transforms.py#L22) calls the `spectrogram` functional which is basically computing the stft and then [getting the power of the complex tensor](https://github.com/pytorch/audio/blob/8273c3f4500ac9c01e95f1d78fa3f5752aafdca7/torchaudio/functional.py#L247) (saying "normalizing" might have confusing on my part, sorry).

But in order to use the phase vocoder, the `TimeStretch` transform has to be applied to the output of the `stft` before the complex norm (i.e. before doing`spec_f = spec_f.pow(power).sum(-1)`).

So that's why I was kinda wanting torchaudio to have a transform that gives the complex `STFT` like we have in contrib, because if someone then wants to work with randomly stretched spectrograms, they can simply do:
```python
nn.Sequential(STFT, TimeStretch, ComplexNorm, ...)
```
whereas currently `STFT` and `ComplexNorm` are "coupled" together as `Spectrogram`, and we can't put the time stretching in between (i.e. `nn.Sequential(Spectrogram, TimeStretch, ...)` won't work).